bee-ledger/src/workers/consensus/white_flag.rs

```diff
@@ -0,0 +1,288 @@
+                balance_diffs.amount_add(*output.address(), output.amount())?;
+                balance_diffs.dust_allowance_add(*output.address(), output.amount())?;
+            }
+            output => return Err(Error::UnsupportedOutputKind(output.kind())),
+        }
+    }
+
+    if created_amount != consumed_amount {
+        return Ok(ConflictReason::InputOutputSumMismatch);
+    }
+
+    for (address, diff) in balance_diffs.iter() {
+        if diff.is_dust_mutating() {
+            let mut balance = storage::fetch_balance_or_default(storage, &address)
+                .await?
+                .apply_diff(diff)?;
+
+            if let Some(diff) = metadata.balance_diffs.get(&address) {
+                balance = balance.apply_diff(diff)?;
+            }
+
+            if balance.dust_outputs() > dust_outputs_max(balance.dust_allowance()) {
+                return Ok(ConflictReason::InvalidDustAllowance);
+            }
+        }
+    }
+
+    for (output_id, created_output) in consumed_outputs {
+        metadata.consumed_outputs.insert(
+            output_id,
+            (created_output, ConsumedOutput::new(*transaction_id, metadata.index)),
+        );
+    }
+
+    for (index, output) in essence.outputs().iter().enumerate() {
+        metadata.created_outputs.insert(
+            OutputId::new(*transaction_id, index as u16)?,
+            CreatedOutput::new(*message_id, output.clone()),
+        );
+    }
+
+    metadata.balance_diffs.merge(balance_diffs)?;
+
+    Ok(ConflictReason::None)
+}
+
+async fn apply_transaction<B: StorageBackend>(
+    storage: &B,
+    message_id: &MessageId,
+    transaction: &TransactionPayload,
+    metadata: &mut WhiteFlagMetadata,
+) -> Result<ConflictReason, Error> {
+    match transaction.essence() {
+        Essence::Regular(essence) => {
+            apply_regular_essence(
+                storage,
+                message_id,
+                &transaction.id(),
+                essence,
+                transaction.unlock_blocks(),
+                metadata,
+            )
+            .await
+        }
+        essence => Err(Error::UnsupportedTransactionEssenceKind(essence.kind())),
+    }
+}
+
+async fn apply_message<B: StorageBackend>(
+    storage: &B,
+    message_id: &MessageId,
+    message: &Message,
+    metadata: &mut WhiteFlagMetadata,
+) -> Result<(), Error> {
+    metadata.referenced_messages += 1;
+
+    match message.payload() {
+        Some(Payload::Transaction(transaction)) => {
+            match apply_transaction(storage, message_id, transaction, metadata).await? {
+                ConflictReason::None => metadata.included_messages.push(*message_id),
+                conflict => metadata.excluded_conflicting_messages.push((*message_id, conflict)),
+            }
+        }
+        _ => metadata.excluded_no_transaction_messages.push(*message_id),
+    }
+
+    Ok(())
+}
+
+async fn traversal<B: StorageBackend>(
```

Comment:
Can we indicate which type of traversal this is to be a bit more self-documenting? Seems like it's traversing the past cone, so maybe `past_cone_traversal`, or just `past_traversal`, or in its imperative form: `traverse_past_cone`/`traverse_past`.